cunumeric/module.py

```diff
@@ -6227,3 +6227,693 @@ def bincount(
+
+# `lower`
+#
+def lower(q, n):
+    gamma = 0.0
+    pos = q * (n - 1)
+    k = math.floor(pos)
+
+    j = int(k)
+    return (gamma, j)
+
+
+# `higher`
+#
+def higher(q, n):
+    pos = q * (n - 1)
+    k = math.floor(pos)
+
+    # Generic rule: (k == pos)
+    gamma = 0.0 if (pos == 0 or k == pos) else 1.0
+
+    j = int(k)
+    return (gamma, j)
+
+
+# `midpoint`
+#
+def midpoint(q, n):
+    pos = q * (n - 1)
+    k = math.floor(pos)
+
+    # Generic rule: (k == pos)
+    gamma = 0.0 if (pos == 0 or k == pos) else 0.5
+
+    j = int(k)
+    return (gamma, j)
+
+
+# `nearest`
+#
+def nearest(q, n):
+    pos = q * (n - 1)
+
+    # k = floor(pos)
+    # gamma = 1.0 if pos - k >= 0.5 else 0.0
+
+    k = np.round(pos)
+    gamma = 0.0
+
+    j = int(k)
+    return (gamma, j)
+
+
+# helper: test if array inputs are different:
+#
+def is_diff(arr1, arr2):
+    return len(arr1) != len(arr2) or any(
+        [x != y for (x, y) in zip(arr1, arr2)]
+    )
+
+
+# for the case when axis = tuple (non-singleton)
+# reshuffling might have to be done (if tuple is non-consecutive)
+# and the src array must be collapsed along that set of axes
+#
+# args:
+#
+# arr:    [in] source nd-array on which quantiles are calculated;
+# axes_set: [in] tuple or list of axes (indices less than arr dimension);
+#
+# return: pair: (minimal_index, reshuffled_and_collapsed source array)
+# TODO: check if reshuffling, reshaping is done in-place!
+#
+def reshuffle_reshape(arr, axes_set):
+    ndim = len(arr.shape)
+
+    sorted_axes = tuple(sorted(axes_set))
+
+    min_dim_index = sorted_axes[0]
+    num_axes = len(sorted_axes)
+    reshuffled_axes = tuple(range(min_dim_index, min_dim_index + num_axes))
+
+    non_consecutive = is_diff(sorted_axes, reshuffled_axes)
+    if non_consecutive:
+        arr_shuffled = moveaxis(arr, sorted_axes, reshuffled_axes)
+    else:
+        arr_shuffled = arr
+
+    # shape_reshuffled = arr_shuffled.shape # debug
+    collapsed_shape = np.product(
+        [arr_shuffled.shape[i] for i in reshuffled_axes]
+    )
+
+    redimed = tuple(range(0, min_dim_index + 1)) + tuple(
+        range(min_dim_index + num_axes, ndim)
+    )
+    reshaped = tuple(
+        [
+            collapsed_shape if k == min_dim_index else arr_shuffled.shape[k]
+            for k in redimed
+        ]
+    )
+
+    arr_reshaped = arr_shuffled.reshape(reshaped)
+    return (min_dim_index, arr_reshaped)
+
+
+# args:
+#
+# arr:      [in] source nd-array on which quantiles are calculated;
+#                preccondition: assumed sorted!
+# q_arr:    [in] quantile input values nd-array;
+# axis:     [in] axis along which quantiles are calculated;
+# method:   [in] func(q, n) returning (gamma, j),
+#                where = array1D.size;
+# keepdims: [in] boolean flag specifying whether collapsed axis
+#                should be kept as dim=1;
+# to_dtype: [in] dtype to convert the result to;
+# qs_all:   [in/out] result pass through or created (returned)
+#
+def quantile_impl(
```

Comment:
Take a look at fixing the errors in the `mypy` check in CI. Most of them are about adding types to your function signatures. You can do the same check locally with `mypy cunumeric/module.py`.

Reply:
Done.